tests/test_dsl.py

```diff
@@ -6,13 +6,30 @@
 from databuilder.concepts import tables
 from databuilder.definition import register
 from databuilder.definition.base import cohort_registry
-from databuilder.dsl import Cohort
-from databuilder.query_language import table
+from databuilder.dsl import (
+    BoolSeries,
+    CodeSeries,
+    Cohort,
+    DateSeries,
+    IntSeries,
+    PatientSeries,
+)
+from databuilder.query_language import Comparator, Value, ValueFromRow, table
 from databuilder.query_utils import get_column_definitions
 
 from .lib.util import OldCohortWithPopulation, make_codelist, mock_positive_tests
 
 
+def a_value():
+    """
+    Build a Value instance
+
+    We use these in various tests, but the actual mechanics of creating this is
+    unimportant to tests themselves.
+    """
+    return Value()
```

Comment:
This is missing my point -- `Value()` is never a valid construction (and so nor is eg `BoolSeries(a_value())`) so we're testing a situation that's never going to be hit in real life.

Perhaps a call tomorrow morning?

Reply:
Sure, is there anything I can look at in the mean time to find what a valid construction should be?

Reply:
Our goals are to ensure that:

* the tests for the DSL only use valid DSL objects (so that we're only testing with realistic values)
* we limit the amount of noise in the tests (so that they're readable)

I originally suggested something like:

```
def test_and():
    series1 = BoolSeries(ValueFromRow(source=None, column="test1"))
    series2 = BoolSeries(ValueFromRow(source=None, column="test2"))

    exp_value = Comparator(
        connector='and_', 
        lhs=Comparator(lhs=series1),
        rhs=Comparator(lhs=series2),
    )

    assert series1 & series2 == BoolSeries(value=exp_value)
```

But this doesn't meet either goal: `ValueFromRow` is not a DSL object, and there's a bit too much detail about the internal structure of `ValueFromRow()`.

Ben's proposal was to have a function that builds some kind of Query Model `Value`, and to use that when building a `BoolSeries` in a test:

```
def a_value():
    ValueFromRow(source=None, column="test1")

def test_and():
    series1 = BoolSeries(a_value())
    series2 = BoolSeries(a_value())

    exp_value = Comparator(
        connector='and_', 
        lhs=Comparator(lhs=series1),
        rhs=Comparator(lhs=series2),
    )

    assert series1 & series2 == BoolSeries(value=exp_value)
```

This makes the test more readable, but `a_value` returns an invalid Query Model object, and so the test is not using a valid DSL object.

The only way to be sure to have a valid DSL object is through using the to DSL.  So we could have one or more functions that build valid DSL objects, by using the DSL:

```
def build_bool_series():
    return clinical_events.exists_for_patient()  # This is about the simplest way to get a valid BoolSeries with the DSL

def test_and():
    series1 = build_bool_series()
    series2 = build_bool_series()

    exp_value = Comparator(
        connector='and_', 
        lhs=Comparator(lhs=series1),
        rhs=Comparator(lhs=series2),
    )

    assert series1 & series2 == BoolSeries(value=exp_value)
```

(`build_bool_series` might be better as a fixture.)

Reply:
There is a problem with the assertions here, which I discovered accidentally because I'm using this branch temporarily for the date arithmetic ticket.  A `BoolSeries` (or any other Series) has had its `__eq__` method overloaded, so the assertion is just asserting a `Comparator`.  
```
assert series1 & series2 == BoolSeries(value=exp_value)
```
is equivalent to 
```
assert Comparator(...)
```
which is always true.

You could change that line to 
```
assert series1 & series2 == "foo"
```
and the test still passes